src/Graph.ts

```diff
@@ -28,14 +28,25 @@ import { Node } from './ast/nodes/shared/Node';
 import Chunk from './Chunk';
 import * as path from './utils/path';
 import GlobalScope from './ast/scopes/GlobalScope';
-import { randomUint8Array, Uint8ArrayXor, Uint8ArrayToHexString } from './utils/entryHashing';
+import { randomUint8Array, Uint8ArrayXor, Uint8ArrayToHexString, isZero } from './utils/entryHashing';
 import { blank } from './utils/object';
 import firstSync from './utils/first-sync';
+import commondir from './utils/commondir';
 
 export type ResolveDynamicImportHandler = (specifier: string | Node, parentId: string) => Promise<string | void>;
 
-function generateChunkName (id: string, chunkNames: { [name: string ]: boolean }, startAtTwo = false): string {
-	let name = path.basename(id);
+function generateChunkName (
+	id: string,
+	chunkNames: { [name: string ]: boolean },
+	startAtTwo = false,
+	inputRelativeDir = ''
+): string {
+	let name;
+	if (inputRelativeDir) {
+		name = path.relative(inputRelativeDir, id).replace(/\\/g, '/');
+	} else {
+		name = path.basename(id);
+	}
```

Comment:
Check indentation here.

Reply:
fixed

Reply:
One thing I observed that I think we should definitely improve is the fact that if a module imports something from outside the common directory of the entry points (i.e. if the entry point imports something that starts with `../`), new files and folders are created outside the intended target directory. This can definitely leave a mess on an unsuspecting user's system. In my opinion there are two possible solutions
1. throw an error if something like this would happen (probably not ideal), or
2. find the common root directory not among the entry points but among the generated chunk file names.

To achieve the latter, we could do the following:
* use `generateChunkName` only for the non-path-preserving case (and remove the last argument)
* if we preserve paths/modules, we use the completed `chunks` object to build a new object with correct paths (possibly extract a helper method here)

This is just a rough idea, some fine-tuning may be necessary.